### ISSUE TITLE
Support for GCMC integration?

### ISSUE THREAD

**Issue by zhenyuwei99** [openmm/openmm#3783]
I am trying to conduct a grand canonical simulation with openmm. But I didn't find relevant information in the doc. I am wondering if there is supported by Openmm?

**Comment by peastman** [comment]
There's no built in support, but it's possible to implement it yourself.  It involves jumping through some hoops though.  What do you want to vary the amount of?  Just water, or something else?

Once you create a Context, the set of particles is fixed.  To change it, you either need to reinitialize it (not hard but a bit expensive) or create a new Context.  Adding particles is easy: just call `addParticle()` on the System and the appropriate methods on all the forces it contains to set the energy terms (`addParticle()`, `addBond()`, etc.), then finish with a call to `reinitialize()` on the Context.  But there's no way to remove particles from a System, so you would have to recreate it from scratch.

What you can do, though, is set the parameters so some particles don't interact with anything.  With a water molecule, for example, if you set `charge` and `epsilon` to 0 for each of the atoms, it will no longer interact with anything else in the system.  The easiest way to do that is calling `setParticleParameters()` on the NonbondedForce for each particle, then `updateParametersInContext()` when you're done to copy the changes over to the Context.

Then of course you need to write your own code for deciding when to change the number of particles.  And when the number increases, you need to somehow pick a position for it.  And does it appear instantly, or gradually fade in to let the rest of the system adapt?

I don't know how much demand there is for grand canonical simulations.  It comes up occasionally but not very often.  We could consider creating a standard implementation to automate a lot of this, at least for a few common cases.

**Comment by peastman** [comment]
Particle indices appear in many places.  For example, many forces list bonds that refer to particles by index.  Likewise for constraints.  And indices within the System need to match indices in the Topology.  If you deleted a particle, then all indices for that particle and ones that came after it would become invalid, and all references to them would no longer be correct.  Trying to manage that would be really complicated and very likely to lead to errors.  That's why it doesn't allow removing particles.

**Comment by zhenyuwei99** [comment]
> would be really complicated and very likely to lead to errors
That's true. I believe the remaining particles with zero parameters only behave like dummy particles, while they still occupy the computational resource.

**Comment by jchodera** [comment]
You can find an example of GCMC implemented within OpenMM (for salt pairs) here---both method and code:
https://www.choderalab.org/publications/2017/11/29/biomolecular-simulations-under-realistic-macroscopic-salt-conditions

The Essex lab also has an OpenMM-based GCMC code for water sampling:
https://github.com/essex-lab/grand

**Comment by zhenyuwei99** [comment]
Really helpful resource. Thanks a lot!
